lib/ruby_smb/gss/provider/ntlm.rb

```diff
@@ -0,0 +1,337 @@
+
+          uint8  :major
+          uint8  :minor
+          uint16 :build
+          uint32 :ntlm_revision, initial_value: 15
+
+          def to_s
+            "Version #{major}.#{minor} (Build #{build}); NTLM Current Revision #{ntlm_revision}"
+          end
+        end
+
+        class Authenticator < Authenticator::Base
+          NEGOTIATE_FLAGS = {
+            :UNICODE                  => 1 << 0,
+            :OEM                      => 1 << 1,
+            :REQUEST_TARGET           => 1 << 2,
+            :SIGN                     => 1 << 4,
+            :SEAL                     => 1 << 5,
+            :DATAGRAM                 => 1 << 6,
+            :LAN_MANAGER_KEY          => 1 << 7,
+            :NTLM                     => 1 << 9,
+            :NT_ONLY                  => 1 << 10,
+            :ANONYMOUS                => 1 << 11,
+            :OEM_DOMAIN_SUPPLIED      => 1 << 12,
+            :OEM_WORKSTATION_SUPPLIED => 1 << 13,
+            :ALWAYS_SIGN              => 1 << 15,
+            :TARGET_TYPE_DOMAIN       => 1 << 16,
+            :TARGET_TYPE_SERVER       => 1 << 17,
+            :TARGET_TYPE_SHARE        => 1 << 18,
+            :EXTENDED_SECURITY        => 1 << 19,
+            :IDENTIFY                 => 1 << 20,
+            :NON_NT_SESSION           => 1 << 22,
+            :TARGET_INFO              => 1 << 23,
+            :VERSION_INFO             => 1 << 25,
+            :KEY128                   => 1 << 29,
+            :KEY_EXCHANGE             => 1 << 30,
+            :KEY56                    => 1 << 31
+          }.freeze
+
+          def reset!
+            super
+            @server_challenge = nil
+          end
+
+          def process(request_buffer=nil)
+            if request_buffer.nil?
+              # this is only NTLMSSP (as opposed to SPNEGO + NTLMSSP)
+              buffer = OpenSSL::ASN1::ASN1Data.new([
+                Gss::OID_SPNEGO,
+                OpenSSL::ASN1::ASN1Data.new([
+                  OpenSSL::ASN1::Sequence.new([
+                    OpenSSL::ASN1::ASN1Data.new([
+                      OpenSSL::ASN1::Sequence.new([
+                        Gss::OID_NTLMSSP
+                      ])
+                    ], 0, :CONTEXT_SPECIFIC),
+                    OpenSSL::ASN1::ASN1Data.new([
+                      OpenSSL::ASN1::ASN1Data.new([
+                        OpenSSL::ASN1::ASN1Data.new([
+                          OpenSSL::ASN1::GeneralString.new('not_defined_in_RFC4178@please_ignore')
+                        ], 0, :CONTEXT_SPECIFIC)
+                      ], 16, :UNIVERSAL)
+                    ], 3, :CONTEXT_SPECIFIC)
+                  ])
+                ], 0, :CONTEXT_SPECIFIC)
+              ], 0, :APPLICATION).to_der
+              return Result.new(buffer, WindowsError::NTStatus::STATUS_SUCCESS)
+            end
+
+            begin
+              gss_api = OpenSSL::ASN1.decode(request_buffer)
+            rescue OpenSSL::ASN1::ASN1Error
+              return
```

Comment:
Yup, I looked around a couple of times to see if we had some kind of logging capability that I could use because you're right. There's a few instances where some kinda error like this will result in the connection being closed.

Reply:
Since this is effectively analogous to Rex::Proto::SMB, can we implement logging a la Rex?

Reply:
Yeah I think that's the way to go. I'll see about adding that on the next major dev pass.